File: SwrveSDKTest/src/test/java/com/swrve/sdk/test/UnsupportedTest.java
package com.swrve.sdk.test;

import android.os.Build;

import com.swrve.sdk.ISwrve;
import com.swrve.sdk.SwrveBaseEmpty;
import com.swrve.sdk.SwrveBaseTest;
import com.swrve.sdk.SwrveHelper;
import com.swrve.sdk.SwrveIAPRewards;
import com.swrve.sdk.SwrveResourceManager;
import com.swrve.sdk.SwrveSDK;
import com.swrve.sdk.SwrveUserResourcesDiffListener;
import com.swrve.sdk.SwrveUserResourcesListener;
import com.swrve.sdk.config.SwrveConfigBase;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.robolectric.util.ReflectionHelpers;

import java.io.File;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

/**
 * Test that the SDK fails gracefully in unsupported versions (lower than 4.X)
 */
public class UnsupportedTest extends SwrveBaseTest {

    private int originalSDKVersion;

    @Before
    public void setUp() throws Exception {
        super.setUp();
        originalSDKVersion = android.os.Build.VERSION.SDK_INT;
        // Explicitly hack the static field that describes the sdk version
        ReflectionHelpers.setStaticField(Build.VERSION.class, "SDK_INT", 15);
    }

    @After
    public void tearDown() throws Exception {
        super.tearDown();
        // revert to the original sdk version
        ReflectionHelpers.setStaticField(Build.VERSION.class, "SDK_INT", originalSDKVersion);
    }

    @Test
    public void testSDKDoesNothingAndDoesntCrash() throws JSONException {

        if (SwrveHelper.sdkAvailable()) {
            // This test should only be executed in platforms levels that the sdk does not support. Otherwise fail it.
            fail("UnsupportedTestHelper explicitly tests sdk on devices where it is not meant to run");
        }

        // Add here all the external API calls and objects that a customer can invoke
        ISwrve sdk = SwrveSDK.createInstance(mActivity.getApplication(), 572, "fake_api_key");
        // The SDK is an empty SDK
        assertTrue(sdk instanceof SwrveBaseEmpty);

        Map<String, String> payload = new HashMap<>();
        payload.put("key", "value");
        SwrveIAPRewards rewards = new SwrveIAPRewards("USD", 99);
        rewards.addCurrency("USD", 100);
        rewards.addItem("stars", 1);
        JSONObject rewardsJson = rewards.getRewardsJSON();
        assertNotNull(rewardsJson);

        sdk.sessionStart();
        sdk.event("new_event");
        sdk.event("event", payload);
        sdk.purchase("item", "USB", 10, 1);
        sdk.currencyGiven("diamonds", 99);
        sdk.userUpdate(payload);
        sdk.iap(1, "productId", 0.99, "USB");
        sdk.iap(1, "productId", 0.99, "USB", rewards);
        SwrveResourceManager resourceManager = sdk.getResourceManager();
        assertNotNull(rewardsJson);

        sdk.setResourcesListener(() -> {
        });
        sdk.getUserResources(new SwrveUserResourcesListener() {
            @Override
            public void onUserResourcesSuccess(Map<String, Map<String, String>> resources, String resourcesAsJSON) {
            }

            @Override
            public void onUserResourcesError(Exception exception) {
            }
        });
        sdk.getUserResourcesDiff(new SwrveUserResourcesDiffListener() {
            @Override
            public void onUserResourcesDiffSuccess(Map<String, Map<String, String>> oldResourcesValues, Map<String, Map<String, String>> newResourcesValues, String resourcesAsJSON) {
            }

            @Override
            public void onUserResourcesDiffError(Exception exception) {
            }
        });
        sdk.sendQueuedEvents();
        sdk.flushToDisk();
        sdk.shutdown();
        sdk.setLanguage(Locale.JAPAN);
        String language = sdk.getLanguage();
        assertEquals("ja-JP", language);

        String apiKey = sdk.getApiKey();
        assertNotNull(apiKey);

        String userId = sdk.getUserId();
        assertNotNull(userId);

        JSONObject deviceInfo = sdk.getDeviceInfo();
        assertNotNull(deviceInfo);

        sdk.refreshCampaignsAndResources();

        String appStoreUrl = sdk.getAppStoreURLForApp(572);
        assertNull(appStoreUrl);

        File cacheDir = sdk.getCacheDir();
        assertNotNull(cacheDir);

        Date initialisedTime = sdk.getInitialisedTime();
        assertNotNull(initialisedTime);

        SwrveConfigBase config = sdk.getConfig();
        assertNotNull(config);

        sdk.stopTracking();
    }
}

File: SwrveSDKTest/src/test/java/com/swrve/sdk/test/UnsupportedTest.kt
package com.swrve.sdk.test

import android.os.Build
import com.swrve.sdk.SwrveBaseEmpty
import com.swrve.sdk.SwrveBaseTest
import com.swrve.sdk.SwrveHelper
import com.swrve.sdk.SwrveIAPRewards
import com.swrve.sdk.SwrveSDK
import com.swrve.sdk.SwrveUserResourcesDiffListener
import com.swrve.sdk.SwrveUserResourcesListener
import com.swrve.sdk.config.SwrveConfigBase
import org.json.JSONException
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.robolectric.util.ReflectionHelpers
import java.util.Locale

/**
 * Test that the SDK fails gracefully in unsupported versions (lower than 4.X)
 */
class UnsupportedTest : SwrveBaseTest() {
    private var originalSDKVersion = 0

    @Before
    @Throws(Exception::class)
    override fun setUp() {
        super.setUp()
        originalSDKVersion = Build.VERSION.SDK_INT
        // Explicitly hack the static field that describes the sdk version
        ReflectionHelpers.setStaticField(Build.VERSION::class.java, "SDK_INT", 15)
    }

    @After
    @Throws(Exception::class)
    override fun tearDown() {
        super.tearDown()
        // revert to the original sdk version
        ReflectionHelpers.setStaticField(Build.VERSION::class.java, "SDK_INT", originalSDKVersion)
    }

    @Test
    @Throws(JSONException::class)
    fun testSDKDoesNothingAndDoesntCrash() {
        if (SwrveHelper.sdkAvailable()) {
            // This test should only be executed in platforms levels that the sdk does not support. Otherwise fail it.
            Assert.fail("UnsupportedTestHelper explicitly tests sdk on devices where it is not meant to run")
        }

        // Add here all the external API calls and objects that a customer can invoke
        val sdk = SwrveSDK.createInstance(mActivity!!.application, 572, "fake_api_key")
        // The SDK is an empty SDK
        Assert.assertTrue(sdk is SwrveBaseEmpty<*, *>)

        val payload: MutableMap<String, String> = HashMap()
        payload["key"] = "value"
        val rewards = SwrveIAPRewards("USD", 99)
        rewards.addCurrency("USD", 100)
        rewards.addItem("stars", 1)
        val rewardsJson = rewards.rewardsJSON
        Assert.assertNotNull(rewardsJson)

        sdk.sessionStart()
        sdk.event("new_event")
        sdk.event("event", payload)
        sdk.purchase("item", "USB", 10, 1)
        sdk.currencyGiven("diamonds", 99.0)
        sdk.userUpdate(payload)
        sdk.iap(1, "productId", 0.99, "USB")
        sdk.iap(1, "productId", 0.99, "USB", rewards)
        val resourceManager = sdk.resourceManager
        Assert.assertNotNull(rewardsJson)

        sdk.setResourcesListener {}
        sdk.getUserResources(object : SwrveUserResourcesListener {
            override fun onUserResourcesSuccess(
                resources: Map<String, Map<String, String>>,
                resourcesAsJSON: String?
            ) {
            }

            override fun onUserResourcesError(exception: Exception) {
            }
        })
        sdk.getUserResourcesDiff(object : SwrveUserResourcesDiffListener {
            override fun onUserResourcesDiffSuccess(
                oldResourcesValues: Map<String, Map<String, String>>,
                newResourcesValues: Map<String, Map<String, String>>,
                resourcesAsJSON: String?
            ) {
            }

            override fun onUserResourcesDiffError(exception: Exception) {
            }
        })
        sdk.sendQueuedEvents()
        sdk.flushToDisk()
        sdk.shutdown()
        sdk.setLanguage(Locale.JAPAN)
        val language = sdk.language
        Assert.assertEquals("ja-JP", language)

        val apiKey = sdk.apiKey
        Assert.assertNotNull(apiKey)

        val userId = sdk.userId
        Assert.assertNotNull(userId)

        val deviceInfo = sdk.deviceInfo
        Assert.assertNotNull(deviceInfo)

        sdk.refreshContent(null)

        val cacheDir = sdk.cacheDir
        Assert.assertNotNull(cacheDir)

        val initialisedTime = sdk.initialisedTime
        Assert.assertNotNull(initialisedTime)

        val config: SwrveConfigBase = sdk.config
        Assert.assertNotNull(config)

        sdk.stopTracking()
    }
}
